src/plopp/core/graph.py

```diff
@@ -17,30 +17,51 @@ def _make_graphviz_digraph(*args, **kwargs):
     return Digraph(*args, **kwargs)
 
 
-def _add_graph_edges(dot, node, inventory, hide_views):
+def _walk_graph(start, nodes, edges, views, hide_views):
     label = escape(str(
-        node.func)) + '\nid = ' + node.id if node.name is None else escape(node.name)
-    inventory.append(node.id)
-    dot.node(node.id, label=label)
-    for child in node.children:
-        key = child.id
-        if key not in inventory:
-            dot.edge(node.id, key)
-            _add_graph_edges(dot, child, inventory, hide_views)
-    for parent in node.parents + list(node.kwparents.values()):
+        start.func)) + '\nid = ' + start.id if start.name is None else escape(
+            start.name)
+    nodes[start.id] = label
+    for child in start.children:
+        if start.id not in edges:
+            edges[start.id] = {child.id}
+        else:
+            edges[start.id].add(child.id)
+        _walk_graph(start=child,
+                    nodes=nodes,
+                    edges=edges,
+                    views=views,
+                    hide_views=hide_views)
+    for parent in start.parents + list(start.kwparents.values()):
         key = parent.id
-        if key not in inventory:
-            dot.edge(key, node.id)
-            _add_graph_edges(dot, parent, inventory, hide_views)
+        if key not in nodes:
+            if key not in edges:
+                edges[key] = {start.id}
+            else:
+                edges[key].add(start.id)
+            _walk_graph(start=parent,
+                        nodes=nodes,
+                        edges=edges,
+                        views=views,
+                        hide_views=hide_views)
```

Comment:
Do I understand correctly that this is a depth first traversal to both children and parents? So conceptually this is equivalent to this?
```python
for neighbor in start.neighbors():
    # ordered nodes on edge (a -> b)
    a, b = get_edge(start, neighbor)
    add_edge(a, b)
    walk(neighbor)
```